### PR TITLE
fix Removed existing taxes while adding new taxes

### PR DESCRIPTION
**Pull request by shantanu-kadam-epi**
Asana Link: https://app.asana.com/0/1200358785304462/1200058080958699

Removed existing taxes from tax table when adding new taxes.

image

### CODE REVIEW

File: erpnext/controllers/accounts_controller.py [hunk 1]
```diff
@@ -922,10 +930,22 @@ def get_taxes_and_charges(master_doctype, master_name):
...
 
 		taxes_and_charges.append(tax)
-
 	return taxes_and_charges
 
 
+def remove_duplicate_taxes(doc_taxes):
+	sales_taxes_accounts = frappe.get_all("Sales Taxes and Charges", fields={"account_head"})
+	sales_taxes_names = [data['account_head'] for data in sales_taxes_accounts]
+	if doc_taxes:
```
Comment (rvpasari) [inline erpnext/controllers/accounts_controller.py:939]: please describe the logic for this in comments

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 2]
```diff
@@ -113,7 +113,30 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
...
+					tax = tax.as_dict()
+
+				doc.set("taxes", [])
+				doc.append("taxes", shipping_charge)
+				doc.append("taxes", tax)
+
+			else:
+				doc.append("taxes", shipping_charge)
+
+	def remove_duplicate_shipping_rule(self,doc, doctype, docname):
```
Comment (MyuddinKhatri) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py]: def remove_duplicate_shipping_rule(self, doc, doctype, docname):

**Review by MyuddinKhatri — CHANGES_REQUESTED** [review]

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 3]
```diff
@@ -113,7 +113,31 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
 		else:
 			shipping_charge["tax_amount"] = shipping_amount
 			shipping_charge["description"] = self.label
-			doc.append("taxes", shipping_charge)
+
+			# Check if already tax is presnt & add new shipping taxes
+			if doc.get("taxes"):
+				for i, tax in enumerate(doc.get("taxes")):
+					tax = tax.as_dict()
```
Comment (hrwX) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py:120]: - Are we assuming that only the previously added tax would be relevant and set it aftre clearing the taxes table ?

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 4]
```diff
@@ -113,7 +113,31 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
...
+
+			else:
+				doc.append("taxes", shipping_charge)
+
+	# remove existing shipping rule tax by its account head from taxes table
+	def remove_existing_shipping_rule(self, doc, doctype, docname):
+		if not docname:
+			return
+
+		shipping_rule_accounts = frappe.get_all("Shipping Rule", fields={"account"})
```
Comment (hrwX) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py]: - fields should be a `list`.

File: erpnext/controllers/accounts_controller.py [hunk 5]
```diff
@@ -922,9 +930,21 @@ def get_taxes_and_charges(master_doctype, master_name):
 				del tax[fieldname]
 
 		taxes_and_charges.append(tax)
-
 	return taxes_and_charges
 
+# remove existing Sales Taxes and Charges by its account head from taxes table
+def remove_existing_sales_taxes(doc_taxes):
+	sales_taxes_accounts = frappe.get_all("Sales Taxes and Charges", fields={"account_head"})
```
Comment (hrwX) [inline erpnext/controllers/accounts_controller.py]: - fields should be a `list`.

File: erpnext/controllers/accounts_controller.py [hunk 6]
```diff
@@ -922,9 +930,21 @@ def get_taxes_and_charges(master_doctype, master_name):
...
+	sales_taxes_names = [data['account_head'] for data in sales_taxes_accounts]
+	if doc_taxes:
+		to_remove = [d for d in doc_taxes
+			if (d['account_head'] in sales_taxes_names)]
```
Comment (hrwX) [inline erpnext/controllers/accounts_controller.py]:
```suggestion
		to_remove = [d for d in doc_taxes if d['account_head'] in sales_taxes_names]
```

File: erpnext/public/js/controllers/transaction.js [hunk 7]
```diff
@@ -1589,20 +1589,20 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
...
 					"master_doctype": frappe.meta.get_docfield(this.frm.doc.doctype, "taxes_and_charges",
 						this.frm.doc.name).options,
-					"master_name": this.frm.doc.taxes_and_charges
+					"master_name": this.frm.doc.taxes_and_charges,
+					"doc": this.frm.doc
 				},
 				callback: function(r) {
 					if(!r.exc) {
 						if(me.frm.doc.shipping_rule && me.frm.doc.taxes) {
+							me.frm.doc.taxes =[];
```
Comment (hrwX) [inline erpnext/public/js/controllers/transaction.js]:
```suggestion
							me.frm.doc.taxes = [];
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py (hunk 4 above)
Comment (shantanu-kadam-epi) [reply erpnext/accounts/doctype/shipping_rule/shipping_rule.py]: As per our discussion, we are deleting the records by its account head only, that's why the records fetched by the account head

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py (hunk 3 above)
Comment (shantanu-kadam-epi) [reply erpnext/accounts/doctype/shipping_rule/shipping_rule.py:120]: Yes

Comment (hrwX) [reply erpnext/accounts/doctype/shipping_rule/shipping_rule.py:120]: Why ?

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py (hunk 4 above)
Comment (hrwX) [reply erpnext/accounts/doctype/shipping_rule/shipping_rule.py]: - I am saying fields should be a `list` not `dict`

**Review by hrwX — CHANGES_REQUESTED** [review]

**Comment by shantanu-kadam-epi** [comment]
Raised New PR for the same, so closing this PR

**Pull request closed without merging**